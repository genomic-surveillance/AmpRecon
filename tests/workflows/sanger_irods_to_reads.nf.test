nextflow_workflow {

    name "Test Workflow SANGER_IRODS_TO_READS"
    script "workflows/sanger_irods_to_reads.nf"
    workflow "SANGER_IRODS_TO_READS"

    test("Should run without failures and output channels should be correct") {

        when {
            params {
                irods_manifest = "/nfs/team335/production/resources/pipeline_resources/plasmodium/falciparum/amplicon/datasets/workflow-testing/sanger_irods_to_reads/irods_manifest_file.tsv"
                channel_data_file = "/nfs/team335/production/resources/pipeline_resources/plasmodium/falciparum/amplicon/datasets/workflow-testing/sanger_irods_to_reads/sanger_irods_to_reads_workflow_data.tsv"
                containers_dir = "/lustre/scratch126/gsu/malariagen/production/resources/pipeline_resources/plasmodium/falciparum/amplicon/ampseq_containers/"
            }
            workflow {
                """
                // Read in workflow channel data file
                channel_data = Channel.fromPath(params.channel_data_file, checkIfExists: true).splitCsv(header: true, sep: '\t')

                // SANGER_IRODS_TO_READS workflow input channels
                input[0] = Channel.fromPath(params.irods_manifest)
                input[1] = channel_data.map { row -> tuple(row.reference_file, row.panel_name, row.snp_list) }
                """
            }
        }

        then {
            // Ensure workflow succeeded
            assert workflow.success

            // validate fastq_ch workflow output channel
            with(workflow.out.fastq_ch) {
                int number_of_lanelets = path(params.irods_manifest).readLines().size() - 1
                // test whether a tuple exists for every lanelet
                assert size() == number_of_lanelets

                def md5_list = [
                    "0dd029efe25ef48245e4cfaab0a52ac4",
                    "d0a27bf0449dd6d9869511b916f6d2a6",
                    "c2037440f5fe1f27642df21268f1696d",
                    "51517e13efdcd47bbc65bf64a01c2bc6"
                ]

                // ensure that:
                // 1) each channel tuple is the correct length
                // 2) the FASTQ file in this tuple has an expected md5 hash
                for (int index = 0; index < number_of_lanelets; index++) {
                    assert get(index).size() == 3
                    with(get(index)) {
                        assert md5_list.contains(path(get(1)).md5)
                    }
                }
            }

            // validate file_id_reference_files_ch workflow output channel
            with(workflow.out.file_id_reference_files_ch) {
                int number_of_lanelets = path(params.irods_manifest).readLines().size() - 1
                // test whether a tuple exists for every lanelet
                assert size() == number_of_lanelets

                // test whether each channel tuple is the correct length
                for (int index = 0; index < number_of_lanelets; index++) {
                    assert get(index).size() == 4
                }
            }

            // validate file_id_to_sample_id_ch workflow output channel
            with(workflow.out.file_id_to_sample_id_ch) {
                int number_of_lanelets = path(params.irods_manifest).readLines().size() - 1
                // test whether a tuple exists for every lanelet
                assert size() == number_of_lanelets

                // test whether each channel tuple is the correct length
                for (int index = 0; index < number_of_lanelets; index++) {
                    assert get(index).size() == 2
                }
            }
        }
    }
}
