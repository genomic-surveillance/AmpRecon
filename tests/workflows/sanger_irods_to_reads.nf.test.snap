{
    "Should run without failures and output should be correct": {
        "content": [
            {
                "0": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "47454_1#1.clipped.fastq:md5,0dd029efe25ef48245e4cfaab0a52ac4",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "47454_1#2.clipped.fastq:md5,d0a27bf0449dd6d9869511b916f6d2a6",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "47454_1#365.clipped.fastq:md5,c2037440f5fe1f27642df21268f1696d",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "47454_1#366.clipped.fastq:md5,51517e13efdcd47bbc65bf64a01c2bc6",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0"
                    ]
                ],
                "1": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "PFA_Spec",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5",
                        "Spec_v1.0.annotation.vcf:md5,491fee622f9538c85a82ec516e992a9c"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "PFA_Spec",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5",
                        "Spec_v1.0.annotation.vcf:md5,491fee622f9538c85a82ec516e992a9c"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "PFA_GRC1_v1.0",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0",
                        "Pf_GRC1v1.0.annotation.vcf:md5,529e3362ad5ded7ff1b14e312b060cc5"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "PFA_GRC1_v1.0",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0",
                        "Pf_GRC1v1.0.annotation.vcf:md5,529e3362ad5ded7ff1b14e312b060cc5"
                    ]
                ],
                "2": [
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "SPT66876"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "SPT66896"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "SPT66897"
                    ]
                ],
                "fastq_ch": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "47454_1#1.clipped.fastq:md5,0dd029efe25ef48245e4cfaab0a52ac4",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "47454_1#2.clipped.fastq:md5,d0a27bf0449dd6d9869511b916f6d2a6",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "47454_1#365.clipped.fastq:md5,c2037440f5fe1f27642df21268f1696d",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "47454_1#366.clipped.fastq:md5,51517e13efdcd47bbc65bf64a01c2bc6",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0"
                    ]
                ],
                "file_id_reference_files_ch": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "PFA_Spec",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5",
                        "Spec_v1.0.annotation.vcf:md5,491fee622f9538c85a82ec516e992a9c"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "PFA_Spec",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5",
                        "Spec_v1.0.annotation.vcf:md5,491fee622f9538c85a82ec516e992a9c"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "PFA_GRC1_v1.0",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0",
                        "Pf_GRC1v1.0.annotation.vcf:md5,529e3362ad5ded7ff1b14e312b060cc5"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "PFA_GRC1_v1.0",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0",
                        "Pf_GRC1v1.0.annotation.vcf:md5,529e3362ad5ded7ff1b14e312b060cc5"
                    ]
                ],
                "file_id_to_sample_id_ch": [
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "SPT66876"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "SPT66896"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "SPT66897"
                    ]
                ]
            }
        ],
        "timestamp": "2023-08-21T08:18:22+0000"
    },
    "Should run without failures": {
        "content": [
            {
                "0": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "47454_1#1.clipped.fastq:md5,0dd029efe25ef48245e4cfaab0a52ac4",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "47454_1#2.clipped.fastq:md5,d0a27bf0449dd6d9869511b916f6d2a6",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "47454_1#365.clipped.fastq:md5,c2037440f5fe1f27642df21268f1696d",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "47454_1#366.clipped.fastq:md5,51517e13efdcd47bbc65bf64a01c2bc6",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0"
                    ]
                ],
                "1": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "PFA_Spec",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5",
                        "Spec_v1.0.annotation.vcf:md5,491fee622f9538c85a82ec516e992a9c"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "PFA_Spec",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5",
                        "Spec_v1.0.annotation.vcf:md5,491fee622f9538c85a82ec516e992a9c"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "PFA_GRC1_v1.0",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0",
                        "Pf_GRC1v1.0.annotation.vcf:md5,529e3362ad5ded7ff1b14e312b060cc5"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "PFA_GRC1_v1.0",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0",
                        "Pf_GRC1v1.0.annotation.vcf:md5,529e3362ad5ded7ff1b14e312b060cc5"
                    ]
                ],
                "2": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "SPT66875"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "SPT66876"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "SPT66896"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "SPT66897"
                    ]
                ],
                "fastq_ch": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "47454_1#1.clipped.fastq:md5,0dd029efe25ef48245e4cfaab0a52ac4",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "47454_1#2.clipped.fastq:md5,d0a27bf0449dd6d9869511b916f6d2a6",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "47454_1#365.clipped.fastq:md5,c2037440f5fe1f27642df21268f1696d",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "47454_1#366.clipped.fastq:md5,51517e13efdcd47bbc65bf64a01c2bc6",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0"
                    ]
                ],
                "file_id_reference_files_ch": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "PFA_Spec",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5",
                        "Spec_v1.0.annotation.vcf:md5,491fee622f9538c85a82ec516e992a9c"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "PFA_Spec",
                        "Spec_v1.0.fasta:md5,1eead451b24a7c91d08e13ff9be3e9e5",
                        "Spec_v1.0.annotation.vcf:md5,491fee622f9538c85a82ec516e992a9c"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "PFA_GRC1_v1.0",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0",
                        "Pf_GRC1v1.0.annotation.vcf:md5,529e3362ad5ded7ff1b14e312b060cc5"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "PFA_GRC1_v1.0",
                        "Pf_GRC1v1.0.fasta:md5,47163a2687af02c95bafcd169fc8e4d0",
                        "Pf_GRC1v1.0.annotation.vcf:md5,529e3362ad5ded7ff1b14e312b060cc5"
                    ]
                ],
                "file_id_to_sample_id_ch": [
                    [
                        "47454_1#1_SPT66875_PFA_Spec",
                        "SPT66875"
                    ],
                    [
                        "47454_1#2_SPT66876_PFA_Spec",
                        "SPT66876"
                    ],
                    [
                        "47454_1#365_SPT66896_PFA_GRC1_v1.0",
                        "SPT66896"
                    ],
                    [
                        "47454_1#366_SPT66897_PFA_GRC1_v1.0",
                        "SPT66897"
                    ]
                ]
            }
        ],
        "timestamp": "2023-08-21T08:18:22+0000"
    }
}